packages/multichain/src/middlewares/multichainMethodCallValidator.ts

```diff
@@ -0,0 +1,122 @@
+import type { Schema, ValidationError } from 'jsonschema';
+import { Validator } from 'jsonschema';
+
+const transformError = (
+  error: ValidationError,
+  param: ContentDescriptorObject,
+  got: unknown,
+) => {
+  // if there is a path, add it to the message
+  const message = `${param.name}${
+    error.path.length > 0 ? `.${error.path.join('.')}` : ''
+  } ${error.message}`;
+
+  return {
+    code: -32602, // TODO: could be a different error code or not wrapped in json-rpc error, since this will also be wrapped in a -32602 invalid params error
+    message,
+    data: {
+      param: param.name,
+      path: error.path,
+      schema: error.schema,
+      got,
+    },
+  };
+};
+
+const v = new Validator();
+
+const dereffedPromise = dereferenceDocument(
+  MultiChainOpenRPCDocument as unknown as OpenrpcDocument,
+  makeCustomResolver({}),
+);
+
+/**
+ * Helper that utilizes the Multichain method specifications from `@metamask/api-specs`
+ * to validate the params of a Multichain request.
+ *
+ * @param method - The request's method.
+ * @param params - The request's optional JsonRpcParams object.
+ * @returns an array of error objects for each validation error or an empty array if no errors.
+ */
+const multichainMethodCallValidator = async (
+  method: string,
+  params: JsonRpcParams | undefined,
+) => {
+  const dereffed = await dereffedPromise;
+
+  const methodToCheck = dereffed.methods.find(
+    (m: MethodObject | ReferenceObject) => (m as MethodObject).name === method,
+  ) as MethodObject | undefined;
+
+  if (
+    !methodToCheck ||
+    !isObject(methodToCheck) ||
+    !('params' in methodToCheck)
+  ) {
+    return [
+      {
+        code: -32601,
+        message: 'The method does not exist / is not available.',
+        data: {
+          method,
+        },
+      },
+    ];
+  }
+
+  const errors: JsonRpcError[] = [];
+  for (const param of methodToCheck.params) {
+    if (!isObject(params)) {
+      return [
+        {
+          code: -32602,
+          message: 'Invalid method parameter(s).',
+        },
+      ];
+    }
+    const p = param as ContentDescriptorObject;
+    const paramToCheck = params[p.name];
+
+    const result = v.validate(paramToCheck, p.schema as unknown as Schema, {
+      required: p.required,
+    });
+    if (result.errors) {
+      errors.push(
+        ...result.errors.map((e) => {
+          return transformError(e, p, paramToCheck) as JsonRpcError;
+        }),
+      );
+    }
+  }
+  return errors;
+};
+
+/**
+ * Middleware that validates the params of a Multichain method request
+ * using the specifications from `@metamask/api-specs`.
+ */
+export const multichainMethodCallValidatorMiddleware = createAsyncMiddleware(
+  async (request, _response, next) => {
+    const errors = await multichainMethodCallValidator(
+      request.method,
+      request.params,
+    );
+    if (errors.length > 0) {
+      throw rpcErrors.invalidParams<JsonRpcError[]>({ data: errors });
```

Comment:
I understand the reasoning for wrapping the errors here, but it might be more ergonomic API-wise to not require the unwrapping on the dapp side? 🤔

Reply:
yeah I feel that. Kindof nice that we can batch send a bunch of errors though

Reply:
The dapps don't have to change how they parse errors, the root level error here is still an Invalid params error as it always was, the only difference is the optional `data` property will be set with additional more specific validation errors.

![image](https://github.com/user-attachments/assets/b3f22c36-6bd3-45e5-937a-587e38d930bd)

Reply:
If they want a proper error message that explains the problem beyond "invalid params" they do have to dig deeper into the error. That is the part I am questioning.

Reply:
Playing devil's advocate, isn't this how JSON-RPC errors are supposed to work? Are we doing anything differently here than other places?